src/Debugger/BreakpointStorage/FileBreakpointStorage.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Debugger\BreakpointStorage;
+
+use Google\Cloud\Core\Lock\FlockLock;
+use Google\Cloud\Core\Lock\LockInterface;
+use Google\Cloud\Debugger\Breakpoint;
+use Google\Cloud\Debugger\Debuggee;
+
+/**
+ * This implementation of BreakpointStorageInterface using a local file.
+ */
+class FileBreakpointStorage implements BreakpointStorageInterface
+{
+    const DEFAULT_FILENAME = 'debugger-breakpoints';
+
+    /* @var string */
+    private $filename;
+
+    /* @var LockInterface */
+    private $lock;
+
+    /**
+     * Create a new FileBreakpointStorage instance.
+     *
+     * @param string $filename [optional] The full path to the storage file.
+     *      **Defaults to** a temporary file in the system's temp directory.
+     */
+    public function __construct($filename = null)
+    {
+        $filename = $filename ?: self::DEFAULT_FILENAME;
+        $this->filename = implode(DIRECTORY_SEPARATOR, [sys_get_temp_dir(), $filename]);
+        $this->lock = new FlockLock($filename . '.lock');
+    }
+
+    /**
+     * Save the given debugger breakpoints.
+     *
+     * @param Debuggee $debuggee
+     * @param Breakpoint[] $breakpoints
+     * @return bool
+     */
+    public function save(Debuggee $debuggee, array $breakpoints)
+    {
+        $data = [
+            'debuggeeId' => $debuggee->id(),
+            'breakpoints' => $breakpoints
+        ];
+
+        $success = false;
+
+        // Acquire an exclusive write lock (blocking). There should only be a
+        // single Daemon that can call this.
+        try {
+            $this->lock->acquire();
```

Comment:
[`LockTrait::synchronize()`](https://github.com/GoogleCloudPlatform/google-cloud-php/blob/master/src/Core/Lock/LockTrait.php#L48) may save a little hassle here, but I don't feel strongly about using it or not.

Reply:
Ah, missed that one. I'll have to propagate the locking options to acquire (depending on if we move the lock options to the constructor).

Reply:
In the case blocking is disabled and the lock fails, synchronize would return `null`. This could make it difficult to differentiate between failing to get the lock and the callable simply executing code that returns `void|null`. WDYT about throwing an exception for users to handle in that case?